src/node_api.cc

```diff
@@ -0,0 +1,2465 @@
+        _finalize_data(finalize_data),
+        _finalize_hint(finalize_hint) {
+    if (initial_refcount == 0) {
+      _persistent.SetWeak(
+          this, FinalizeCallback, v8::WeakCallbackType::kParameter);
+      _persistent.MarkIndependent();
+    }
+  }
+
+  ~Reference() {
+    // The V8 Persistent class currently does not reset in its destructor:
+    // see NonCopyablePersistentTraits::kResetInDestructor = false.
+    // (Comments there claim that might change in the future.)
+    // To avoid memory leaks, it is better to reset at this time, however
+    // care must be taken to avoid attempting this after the Isolate has
+    // shut down, for example via a static (atexit) destructor.
+    _persistent.Reset();
+  }
+
+ public:
+  static Reference* New(v8::Isolate* isolate,
+                        v8::Local<v8::Value> value,
+                        int initial_refcount,
+                        bool delete_self,
+                        napi_finalize finalize_callback = nullptr,
+                        void* finalize_data = nullptr,
+                        void* finalize_hint = nullptr) {
+    return new Reference(isolate,
+      value,
+      initial_refcount,
+      delete_self,
+      finalize_callback,
+      finalize_data,
+      finalize_hint);
+  }
+
+  static void Delete(Reference* reference) {
+    delete reference;
+  }
+
+  int Ref() {
+    if (++_refcount == 1) {
+      _persistent.ClearWeak();
+    }
+
+    return _refcount;
+  }
+
+  int Unref() {
+    if (--_refcount == 0) {
+      _persistent.SetWeak(
+          this, FinalizeCallback, v8::WeakCallbackType::kParameter);
+      _persistent.MarkIndependent();
+    }
+
+    return _refcount;
+  }
+
+  v8::Local<v8::Value> Get() {
+    if (_persistent.IsEmpty()) {
+      return v8::Local<v8::Value>();
+    } else {
+      return v8::Local<v8::Value>::New(_isolate, _persistent);
+    }
+  }
+
+ private:
+  static void FinalizeCallback(const v8::WeakCallbackInfo<Reference>& data) {
+    Reference* reference = data.GetParameter();
+    reference->_persistent.Reset();
+
+    // Check before calling the finalize callback, because the callback might
+    // delete it.
+    bool delete_self = reference->_delete_self;
+
+    if (reference->_finalize_callback != nullptr) {
+      reference->_finalize_callback(reference->_finalize_data,
+          reference->_finalize_hint);
+    }
+
+    if (delete_self) {
+      delete reference;
+    }
+  }
+
+  v8::Isolate* _isolate;
+  v8::Persistent<v8::Value> _persistent;
+  int _refcount;
+  bool _delete_self;
+  napi_finalize _finalize_callback;
+  void* _finalize_data;
+  void* _finalize_hint;
+};
+
+class TryCatch : public v8::TryCatch {
+ public:
+  explicit TryCatch(v8::Isolate* isolate)
+      : v8::TryCatch(isolate), _isolate(isolate) {}
+
+  ~TryCatch() {
+    if (HasCaught()) {
+      _the_exception.Reset(_isolate, Exception());
+    }
+  }
+
+  static v8::Persistent<v8::Value>& LastException() { return _the_exception; }
```

Comment:
By the way, I’d be glad if we could avoid globals and instead turn these into per-Isolate/Environment properties… that shouldn’t be a blocker for this PR, though